src/Controller/Api/Legacy/GalleryController.php

```diff
@@ -0,0 +1,520 @@
+    protected function handleWriteGalleryhasmedia(GalleryInterface $gallery, MediaInterface $media, ?GalleryHasMediaInterface $galleryHasMedia = null, Request $request)
+    {
+        $form = $this->formFactory->createNamed(null, ApiGalleryHasMediaType::class, $galleryHasMedia, [
+            'csrf_protection' => false,
+        ]);
+
+        $form->handleRequest($request);
+
+        if ($form->isValid()) {
+            $galleryHasMedia = $form->getData();
+            $galleryHasMedia->setMedia($media);
+
+            // NEXT_MAJOR: remove this if/else block. Just call `$gallery->addGalleryHasMedia($galleryHasMedia);`
+            if ($gallery instanceof GalleryMediaCollectionInterface) {
+                $gallery->addGalleryHasMedia($galleryHasMedia);
+            } else {
+                $gallery->addGalleryHasMedias($galleryHasMedia);
+            }
+            $this->galleryManager->save($gallery);
+
+            $context = new Context();
+            $context->setGroups(['sonata_api_read']);
+            $context->enableMaxDepth();
+
+            $view = FOSRestView::create($galleryHasMedia);
+            $view->setContext($context);
+
+            return $view;
+        }
+
+        return $form;
+    }
+
+    /**
+     * Retrieves gallery with identifier $id or throws an exception if it doesn't exist.
+     *
+     * @param string $id Gallery identifier
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return GalleryInterface
+     */
+    protected function getGallery($id)
+    {
+        $gallery = $this->getGalleryManager()->findOneBy(['id' => $id]);
+
+        if (null === $gallery) {
+            throw new NotFoundHttpException(sprintf('Gallery (%s) not found', $id));
+        }
+
+        return $gallery;
+    }
+
+    /**
+     * Retrieves media with identifier $id or throws an exception if it doesn't exist.
+     *
+     * @param string $id Media identifier
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return MediaInterface
+     */
+    protected function getMedia($id)
+    {
+        $media = $this->getMediaManager()->findOneBy(['id' => $id]);
+
+        if (null === $media) {
+            throw new NotFoundHttpException(sprintf('Media (%s) not found', $id));
+        }
+
+        return $media;
+    }
+
+    /**
+     * @return GalleryManagerInterface
+     */
+    protected function getGalleryManager()
+    {
+        return $this->galleryManager;
+    }
+
+    /**
+     * @return MediaManagerInterface
+     */
+    protected function getMediaManager()
+    {
+        return $this->mediaManager;
+    }
+
+    /**
+     * Write a Gallery, this method is used by both POST and PUT action methods.
+     *
+     * @param Request     $request Symfony request
+     * @param string|null $id      Gallery identifier
+     *
+     * @return Rest\View|FormInterface
+     */
+    protected function handleWriteGallery($request, $id = null)
+    {
+        $gallery = $id ? $this->getGallery($id) : null;
```

Comment:
Although it could considered as a border case, I think `'0'` is a valid id.
```suggestion
        $gallery = null !== $id ? $this->getGallery($id) : null;
```

Reply:
`$this->getGallery($id)` must return `GalleryInterface`, so:
```suggestion
        $gallery = $this->getGallery($id);
```